// sb-exec/lib/index.js
'use strict';

var exec = function () {
  var _ref = _asyncToGenerator(function* (givenFilePath, givenParameters, options, callback) {
    var nodeSpawnOptions = yield (0, _helpers.getSpawnOptions)(options);
    var filePath = givenFilePath;
    var parameters = givenParameters;
    var spawnedCmdOnWindows = false;

    if ((0, _helpers.shouldNormalizeForWindows)(filePath, options)) {
      nodeSpawnOptions.windowsVerbatimArguments = true;
      var cmdArgs = [filePath];
      // filePath must be escaped if it has a \s in it, otherwise it must not
      if (/\s/.test(filePath)) {
        cmdArgs = cmdArgs.map(_helpers.escape);
      }
      cmdArgs = cmdArgs.concat(parameters.map(_helpers.escape));
      filePath = process.env.comspec || 'cmd.exe';
      parameters = ['/s', '/c', '"' + cmdArgs.join(' ') + '"'];
      spawnedCmdOnWindows = true;
    }

    return new Promise(function (resolve, reject) {
      var spawnedProcess = (0, _child_process.spawn)(filePath, parameters, nodeSpawnOptions);
      var data = { stdout: [], stderr: [] };
      var timeout = void 0;

      if (spawnedProcess.stdout) {
        spawnedProcess.stdout.on('data', function (chunk) {
          data.stdout.push(chunk);
        });
      }
      if (spawnedProcess.stderr) {
        spawnedProcess.stderr.on('data', function (chunk) {
          data.stderr.push(chunk);
        });
      }
      spawnedProcess.on('error', function (error) {
        reject(error);
      });
      spawnedProcess.on('close', function (exitCode) {
        clearTimeout(timeout);

        var stdout = data.stdout.join('').trim();
        var stderr = data.stderr.join('').trim();

        // NOTE: On windows, we spawn everything through cmd.exe
        // So we have to manually construct ENOENT from it's error message
        if (spawnedCmdOnWindows && stderr === '\'' + givenFilePath + '\' is not recognized as an internal or external command,\r\noperable program or batch file.') {
          reject((0, _helpers.getENOENTError)(givenFilePath, givenParameters));
          return;
        }

        if (options.stream === 'stdout') {
          if (stderr && options.throwOnStderr) {
            reject(new Error(stderr));
          } else if (exitCode !== 0 && !options.ignoreExitCode) {
            console.error('[exec] Process exited with no-zero code, stdout: ', stdout);
            reject(new Error('Process exited with non-zero code: ' + exitCode));
          } else {
            resolve(stdout);
          }
        } else if (options.stream === 'stderr') {
          if (stderr.length === 0 && !options.allowEmptyStderr) {
            reject(new Error('Process exited with no output, code: ' + exitCode));
          } else {
            resolve(stderr);
          }
        } else {
          resolve({ stdout: stdout, stderr: stderr, exitCode: exitCode });
        }
      });

      if (spawnedProcess.stdin) {
        if (options.stdin) {
          try {
            spawnedProcess.stdin.write(options.stdin);
          } catch (_) {/* No Op */}
        }
        try {
          spawnedProcess.stdin.end();
        } catch (_) {/* No Op */}
      }

      if (options.timeout !== Infinity) {
        timeout = setTimeout(function () {
          // eslint-disable-next-line no-use-before-define
          killProcess(spawnedProcess);
          reject(new Error('Process execution timed out'));
        }, options.timeout);
      }
      callback(spawnedProcess);
    });
  });

  return function exec(_x, _x2, _x3, _x4) {
    return _ref.apply(this, arguments);
  };
}();

// NOTE: This proxy function is required to allow .kill() in different stages of process spawn
// We cannot put this logic into exec() directly because it's an async function and doesn't
// allow us to temper the underlying promise


var killProcess = function () {
  var _ref2 = _asyncToGenerator(function* (spawnedProcess) {
    var signal = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'SIGTERM';

    var spawnfile = _path2.default.basename(spawnedProcess.spawnfile);
    if (process.platform !== 'win32' || spawnfile === 'wmic.exe' || spawnfile === 'wmic') {
      // Also do this if the process is wmic
      spawnedProcess.kill(signal);
      return;
    }
    try {
      var output = yield execProxy('wmic', ['process', 'where', '(ParentProcessId=' + spawnedProcess.pid + ')', 'get', 'processid'], { stream: 'stdout', timeout: 60 * 1000 });
      output.split(/\s+/).filter(function (i) {
        return (/^\d+$/.test(i)
        );
      }).map(function (i) {
        return parseInt(i, 10);
      }).filter(function (i) {
        return i !== spawnedProcess.pid && i > 0;
      }).forEach(function (pid) {
        process.kill(pid, signal);
      });
    } catch (error) {/* No Op */}
  });

  return function killProcess(_x7) {
    return _ref2.apply(this, arguments);
  };
}();

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _child_process = require('child_process');

var _helpers = require('./helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function execProxy(filePath) {
  var parameters = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var givenOptions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  var options = (0, _helpers.validate)(filePath, parameters, givenOptions);
  var killSignal = null;
  var spawnedProcess = null;
  var promise = exec(filePath, parameters, options, function (spawnedChildProcess) {
    if (killSignal) {
      // eslint-disable-next-line no-use-before-define
      killProcess(spawnedChildProcess, killSignal);
    } else {
      spawnedProcess = spawnedChildProcess;
    }
  });
  promise.kill = function (givenKillSignal) {
    if (spawnedProcess) {
      // eslint-disable-next-line no-use-before-define
      killProcess(spawnedProcess, givenKillSignal);
    } else {
      killSignal = givenKillSignal || 'SIGTERM';
    }
  };
  return promise;
}

function execNode(filePath) {
  var parameters = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var givenOptions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  (0, _helpers.validate)(filePath, parameters, givenOptions);
  return execProxy(process.execPath, [filePath].concat(parameters), givenOptions);
}

module.exports = { exec: execProxy, execNode: execNode };
