// solid-js/web/dist/server.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var stream = require('stream');
var solidJs = require('solid-js');

const booleans = ["allowfullscreen", "allowpaymentrequest", "async", "autofocus", "autoplay", "checked", "controls", "default", "disabled", "formnovalidate", "hidden", "ismap", "itemscope", "loop", "multiple", "muted", "nomodule", "novalidate", "open", "playsinline", "readonly", "required", "reversed", "seamless", "selected", "truespeed"];
const BooleanAttributes = new Set(booleans);
const Aliases = {
  className: "class",
  htmlFor: "for"
};

function memo(fn, equal) {
  return solidJs.createMemo(fn, undefined, equal);
}

function dynamicProperty(props, key) {
  const src = props[key];
  Object.defineProperty(props, key, {
    get() {
      return src();
    },
    enumerable: true
  });
  return props;
}
function getHydrationKey() {
  return globalThis._$HYDRATION.context.id;
}

function ssrClassList(value) {
  let classKeys = Object.keys(value),
      result = "";
  for (let i = 0, len = classKeys.length; i < len; i++) {
    const key = classKeys[i],
          classValue = !!value[key];
    if (!key || !classValue) continue;
    i && (result += " ");
    result += key;
  }
  return result;
}
function ssrStyle(value) {
  if (typeof value === "string") return value;
  let result = "";
  const k = Object.keys(value);
  for (let i = 0; i < k.length; i++) {
    const s = k[i];
    if (i) result += ";";
    result += `${s}:${escape(value[s], true)}`;
  }
  return result;
}
function ssrSpread(props) {
  return () => {
    if (typeof props === "function") props = props();
    const keys = Object.keys(props);
    let result = "";
    for (let i = 0; i < keys.length; i++) {
      const prop = keys[i];
      if (prop === "children") {
        console.warn(`SSR currently does not support spread children.`);
        continue;
      }
      const value = props[prop];
      if (prop === "style") {
        result += `style="${ssrStyle(value)}"`;
      } else if (prop === "classList") {
        result += `class="${ssrClassList(value)}"`;
      } else if (BooleanAttributes.has(prop)) {
        if (value) result += prop;else continue;
      } else {
        result += `${Aliases[prop] || prop}="${escape(value, true)}"`;
      }
      if (i !== keys.length - 1) result += " ";
    }
    return result;
  };
}
function ssrBoolean(key, value) {
  return value ? " " + key : "";
}
function escape(s, attr) {
  const t = typeof s;
  if (t !== "string") {
    if (attr && t === "boolean") return String(s);
    return s;
  }
  const delim = attr ? '"' : "<";
  const escDelim = attr ? "&quot;" : "&lt;";
  let iDelim = s.indexOf(delim);
  let iAmp = s.indexOf("&");
  if (iDelim < 0 && iAmp < 0) return s;
  let left = 0,
      out = "";
  while (iDelim >= 0 && iAmp >= 0) {
    if (iDelim < iAmp) {
      if (left < iDelim) out += s.substring(left, iDelim);
      out += escDelim;
      left = iDelim + 1;
      iDelim = s.indexOf(delim, left);
    } else {
      if (left < iAmp) out += s.substring(left, iAmp);
      out += "&amp;";
      left = iAmp + 1;
      iAmp = s.indexOf("&", left);
    }
  }
  if (iDelim >= 0) {
    do {
      if (left < iDelim) out += s.substring(left, iDelim);
      out += escDelim;
      left = iDelim + 1;
      iDelim = s.indexOf(delim, left);
    } while (iDelim >= 0);
  } else while (iAmp >= 0) {
    if (left < iAmp) out += s.substring(left, iAmp);
    out += "&amp;";
    left = iAmp + 1;
    iAmp = s.indexOf("&", left);
  }
  return left < s.length ? out + s.substring(left) : out;
}
function resolveSSRNode(node) {
  const t = typeof node;
  if (t === "string") return node;
  if (node == null || t === "boolean") return "";
  if (Array.isArray(node)) return node.map(resolveSSRNode).join("");
  if (t === "object") return resolveSSRNode(node.t);
  if (t === "function") return resolveSSRNode(node());
  return String(node);
}

function renderToNodeStream(code) {
  const stream$1 = new stream.Readable({
    read() {}
  });
  const hydration = globalThis._$HYDRATION || (globalThis._$HYDRATION = {});
  hydration.context = {
    id: "0",
    count: 0
  };
  hydration.streamSSR = true;
  let count = 0,
      completed = 0,
      checkEnd = () => {
    if (completed === count) {
      stream$1.push(null);
      delete hydration.context;
    }
  };
  hydration.register = p => {
    const id = ++count;
    p.then(d => {
      stream$1.push(`<script>_$HYDRATION.resolveResource(${id}, ${JSON.stringify(d)})</script>`);
      ++completed && checkEnd();
    });
  };
  stream$1.push(resolveSSRNode(code()));
  setTimeout(checkEnd);
  return stream$1;
}
function renderToWebStream(code) {
  const {
    readable,
    writable
  } = new TransformStream();
  const writer = writable.getWriter();
  const encoder = new TextEncoder();
  const hydration = globalThis._$HYDRATION || (globalThis._$HYDRATION = {});
  hydration.context = {
    id: "0",
    count: 0
  };
  hydration.streamSSR = true;
  let count = 0,
      completed = 0,
      checkEnd = () => {
    if (hydration.context && completed === count) {
      writer.close();
      delete hydration.context;
    }
  };
  hydration.register = p => {
    const id = ++count;
    p.then(d => {
      writer.write(encoder.encode(`<script>_$HYDRATION.resolveResource(${id}, ${JSON.stringify(d).replace(/'/g, "\\'").replace(/\\\"/g, '\\\\\\"')})</script>`));
      ++completed && checkEnd();
    });
  };
  writer.write(encoder.encode(resolveSSRNode(code())));
  setTimeout(checkEnd);
  return readable;
}
function renderToString(code) {
  const hydration = globalThis._$HYDRATION || (globalThis._$HYDRATION = {});
  hydration.context = {
    id: "0",
    count: 0
  };
  return resolveSSRNode(code());
}
function ssr(t, ...nodes) {
  if (nodes.length) {
    let result = "";
    for (let i = 0; i < t.length; i++) {
      result += t[i];
      const node = nodes[i];
      if (node !== undefined) result += resolveSSRNode(node);
    }
    t = result;
  }
  return {
    t
  };
}
function generateHydrationScript({
  eventNames = ["click", "input", "blur"],
  streaming
} = {}) {
  let s = `(()=>{_$HYDRATION={events:[],completed:new WeakSet};const t=e=>e&&e.hasAttribute&&(e.hasAttribute("data-hk")&&e||t(e.host&&e.host instanceof Node?e.host:e.parentNode)),e=e=>{let o=e.composedPath&&e.composedPath()[0]||e.target,s=t(o);s&&!_$HYDRATION.completed.has(s)&&_$HYDRATION.events.push([s,e])};["${eventNames.join('","')}"].forEach(t=>document.addEventListener(t,e))})();`;
  if (streaming) {
    s += `(()=>{const e=_$HYDRATION,r={};let o=0;e.resolveResource=((e,o)=>{const t=r[e];if(!t)return r[e]=o;delete r[e],t(o)}),e.loadResource=(()=>{const e=++o,t=r[e];if(!t){let o,t=new Promise(e=>o=e);return r[e]=o,t}return delete r[e],Promise.resolve(t)})})();`;
  }
  return s;
}

const isServer = true;
function spread() {}
function Dynamic(props) {
  const [p, others] = solidJs.splitProps(props, ["component"]);
  const comp = p.component,
        t = typeof comp;
  if (comp) {
    if (t === "function") return comp(others);else if (t === "string") {
      const [local, sOthers] = solidJs.splitProps(others, ["children"]);
      return ssr([`<${comp} `, ">", `</${comp}>`], ssrSpread(sOthers), local.children || "");
    }
  }
}
function Portal(props) {
  return "";
}

Object.defineProperty(exports, 'ErrorBoundary', {
  enumerable: true,
  get: function () {
    return solidJs.ErrorBoundary;
  }
});
Object.defineProperty(exports, 'For', {
  enumerable: true,
  get: function () {
    return solidJs.For;
  }
});
Object.defineProperty(exports, 'Index', {
  enumerable: true,
  get: function () {
    return solidJs.Index;
  }
});
Object.defineProperty(exports, 'Match', {
  enumerable: true,
  get: function () {
    return solidJs.Match;
  }
});
Object.defineProperty(exports, 'Show', {
  enumerable: true,
  get: function () {
    return solidJs.Show;
  }
});
Object.defineProperty(exports, 'Suspense', {
  enumerable: true,
  get: function () {
    return solidJs.Suspense;
  }
});
Object.defineProperty(exports, 'SuspenseList', {
  enumerable: true,
  get: function () {
    return solidJs.SuspenseList;
  }
});
Object.defineProperty(exports, 'Switch', {
  enumerable: true,
  get: function () {
    return solidJs.Switch;
  }
});
Object.defineProperty(exports, 'assignProps', {
  enumerable: true,
  get: function () {
    return solidJs.assignProps;
  }
});
Object.defineProperty(exports, 'createComponent', {
  enumerable: true,
  get: function () {
    return solidJs.createComponent;
  }
});
Object.defineProperty(exports, 'currentContext', {
  enumerable: true,
  get: function () {
    return solidJs.getContextOwner;
  }
});
Object.defineProperty(exports, 'effect', {
  enumerable: true,
  get: function () {
    return solidJs.createRenderEffect;
  }
});
exports.Dynamic = Dynamic;
exports.Portal = Portal;
exports.dynamicProperty = dynamicProperty;
exports.escape = escape;
exports.generateHydrationScript = generateHydrationScript;
exports.getHydrationKey = getHydrationKey;
exports.isServer = isServer;
exports.memo = memo;
exports.renderToNodeStream = renderToNodeStream;
exports.renderToString = renderToString;
exports.renderToWebStream = renderToWebStream;
exports.resolveSSRNode = resolveSSRNode;
exports.spread = spread;
exports.ssr = ssr;
exports.ssrBoolean = ssrBoolean;
exports.ssrClassList = ssrClassList;
exports.ssrSpread = ssrSpread;
exports.ssrStyle = ssrStyle;
