// atom-utils-plus/lib/mixins/space-pen-dsl.js
// Generated by CoffeeScript 2.5.1
(function() {
  var BabelSpacePenDSL, Builder, Events, Mixin, SelfClosingTags, SpacePenDSL, Tags, Template;

  Mixin = require('mixto');

  Tags = 'a abbr address article aside audio b bdi bdo blockquote body button canvas caption cite code colgroup datalist dd del details dfn dialog div dl dt em fieldset figcaption figure footer form h1 h2 h3 h4 h5 h6 head header html i iframe ins kbd label legend li main map mark menu meter nav noscript object ol optgroup option output p pre progress q rp rt ruby s samp script section select small span strong style sub summary sup table tbody td textarea tfoot th thead time title tr u ul var video area base br col command embed hr img input keygen link meta param source track wbr'.split(/\s+/);

  SelfClosingTags = {};

  'area base br col command embed hr img input keygen link meta param source track wbr'.split(/\s+/).forEach(function(tag) {
    return SelfClosingTags[tag] = true;
  });

  Events = 'blur change click dblclick error focus input keydown keypress keyup load mousedown mousemove mouseout mouseover mouseup resize scroll select submit unload'.split(/\s+/);

  BabelSpacePenDSL = class BabelSpacePenDSL extends Mixin {
    buildContent() {
      if (this.constructor.content != null) {
        return SpacePenDSL.buildContent(this, this.constructor.content);
      }
    }

  };

  module.exports = SpacePenDSL = (function() {
    class SpacePenDSL extends Mixin {
      static includeInto(klass) {
        super.includeInto(klass);
        Object.defineProperty(klass, 'content', {
          enumerable: false,
          get: function() {
            return this.prototype.__content__;
          },
          set: function(value) {
            return this.prototype.__content__ = value;
          }
        });
        Object.defineProperty(klass.prototype, 'createdCallback', {
          enumerable: true,
          get: function() {
            return this.__create__;
          },
          set: function(value) {
            return this.__createdCallback__ = value;
          }
        });
        Object.defineProperty(klass.prototype, '__create__', {
          enumerable: true,
          value: function() {
            if (this.__content__ != null) {
              SpacePenDSL.buildContent(this, this.__content__);
            }
            if (this.__createdCallback__ != null) {
              return this.__createdCallback__();
            }
          }
        });
        return klass.useShadowRoot = function() {
          return klass.prototype.__useShadowRoot__ = true;
        };
      }

      static buildContent(element, content) {
        var html, root, template;
        template = new Template();
        content.call(template);
        [html] = template.buildHtml();
        root = element.__useShadowRoot__ ? root = element.shadowRoot = element.createShadowRoot() : root = element;
        root.innerHTML = html;
        return this.wireOutlets(element, root);
      }

      static wireOutlets(view, root) {
        var element, i, len, outlet, ref;
        ref = root.querySelectorAll('[outlet]');
        for (i = 0, len = ref.length; i < len; i++) {
          element = ref[i];
          outlet = element.getAttribute('outlet');
          view[outlet] = element;
          element.removeAttribute('outlet');
        }
        return void 0;
      }

    };

    SpacePenDSL.Babel = BabelSpacePenDSL;

    return SpacePenDSL;

  }).call(this);

  Template = (function() {
    class Template {
      constructor() {
        this.currentBuilder = new Builder();
      }

      subview(name, view) {
        return this.currentBuilder.subview(name, view);
      }

      text(string) {
        return this.currentBuilder.text(string);
      }

      tag(tagName, ...args) {
        return this.currentBuilder.tag(tagName, ...args);
      }

      raw(string) {
        return this.currentBuilder.raw(string);
      }

      buildHtml() {
        return this.currentBuilder.buildHtml();
      }

    };

    Tags.forEach(function(tagName) {
      return Template.prototype[tagName] = function(...args) {
        return this.currentBuilder.tag(tagName, ...args);
      };
    });

    return Template;

  }).call(this);

  Builder = class Builder {
    constructor() {
      this.document = [];
      this.postProcessingSteps = [];
    }

    buildHtml() {
      return [this.document.join(''), this.postProcessingSteps];
    }

    tag(name, ...args) {
      var options;
      options = this.extractOptions(args);
      this.openTag(name, options.attributes);
      if (SelfClosingTags.hasOwnProperty(name)) {
        if ((options.text != null) || (options.content != null)) {
          throw new Error(`Self-closing tag ${name} cannot have text or content`);
        }
      } else {
        if (typeof options.content === "function") {
          options.content();
        }
        if (options.text) {
          this.text(options.text);
        }
        return this.closeTag(name);
      }
    }

    openTag(name, attributes) {
      var attributeName, attributePairs, attributesString, value;
      if (this.document.length === 0) {
        if (attributes == null) {
          attributes = {};
        }
      }
      attributePairs = (function() {
        var results;
        results = [];
        for (attributeName in attributes) {
          value = attributes[attributeName];
          results.push(`${attributeName}=\"${value}\"`);
        }
        return results;
      })();
      attributesString = attributePairs.length ? " " + attributePairs.join(" ") : "";
      return this.document.push(`<${name}${attributesString}>`);
    }

    closeTag(name) {
      return this.document.push(`</${name}>`);
    }

    text(string) {
      var escapedString;
      escapedString = string.replace(/&/g, '&amp;').replace(/"/g, '&quot;').replace(/'/g, '&#39;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
      return this.document.push(escapedString);
    }

    raw(string) {
      return this.document.push(string);
    }

    subview(outletName, subview) {
      var subviewId;
      subviewId = `subview-${++idCounter}`;
      this.tag('div', {
        id: subviewId
      });
      return this.postProcessingSteps.push(function(view) {
        view[outletName] = subview;
        subview.parentView = view;
        return view.find(`div#${subviewId}`).replaceWith(subview);
      });
    }

    extractOptions(args) {
      var arg, i, len, options;
      options = {};
      for (i = 0, len = args.length; i < len; i++) {
        arg = args[i];
        switch (typeof arg) {
          case 'function':
            options.content = arg;
            break;
          case 'string':
          case 'number':
            options.text = arg.toString();
            break;
          default:
            options.attributes = arg;
        }
      }
      return options;
    }

  };

}).call(this);
